cpp/include/cudf/utilities/type_dispatcher.hpp

```diff
@@ -104,6 +104,51 @@ using device_storage_type_t =
   std::conditional_t<std::is_same_v<numeric::decimal128, T>, __int128_t, T>>>;
 // clang-format on
 
+/**
+ * @brief Maps types with the same width to corresponding integer type.
+ *
+ * Types like `uint32`, `float`, `timestamp_D` and `duration_D` map to `int32`
+ * since all are 4 bytes wide.
+ *
+ * This should only be used in operations that simply copy data in memory and
+ * do no require using the data values within the operation.
+ * For example, `gather`, `concatenate`, `interleave` only need to copy data.
+ * Whereas, `scan`, `sort`, `merge` require the data values.
+ *
+ * The types are enumerated to prevent fixed-point types from incorrectly
+ * mapping to incompatible output types.
+ *
+ * Use this "type function" with the `using` type alias:
+ * @code
+ * using Type = device_storage_width_t<Element>;
+ * @endcode
+ *
+ * @tparam T The literal type that is stored on the host
+ */
+// clang-format off
+template <typename T>
```

Comment:
Yes, the nested `if std::conditional_t` here eventually resolve to just return `T` at the last line if it does not match anything else:
```
std::conditional_t<std::is_same_v<numeric::decimal128, T>, __int128_t, T>>>>>
                                                                       ^
```
This is to cover non-fixed width types in general but will also applies to any future types without error. This is merely a compile-time improvement so adding a new type without including it here would just generate an extra code path in all the copy-ish places that use this dispatch type.

Reply:
That's what I'm concerned about. It would be nice if one passes an unforeseen type that this would fail loudly so we don't forget to add the new types.

Second, the `_t` suffix is reserved by POSIX AFAIK, so this should probably end in `_type`, not `_t`.

Third, this returns a type, not a width, but the name `device_storage_width_t` is a bit confusing because it's unclear if it is a number of bytes or a type.  Also generally "size" is used for byte size, not "width".

Would `device_equal_size_type` or `device_size_proxy_type` or something like that would be clearer?

Reply:
I was using `width` so to be more inline with `is_fixed_width` where `width` is used to identify the number of bytes in the type.

Reply:
> That's what I'm concerned about. It would be nice if one passes an unforeseen type that this would fail loudly so we don't forget to add the new types.

Sorry, I don't really know how to check for unforeseen types not being added here.

Reply:
Neither do I. Which is why I started the discussion. I predict hard to debug, silent failures.  We should at least be aware of this limitation, and perhaps document it.

I'm OK with "width", but I think the name should be clarified so it's not unclear whether it results in a width or a type.

Reply:
> I predict hard to debug, silent failures. We should at least be aware of this limitation, and perhaps document it.

Sorry, I'm confused. What kind of failures are you referencing?